Implement generics syntax to allow improved static analysis

Mark-H opened the pull request:
I've been playing with phpstan the last few days and because xPDO is everywhere, this is something I've been running into. getObject()/getCollection() etc returning an xPDOObject is not specific enough.

These docblock changes allow phpstan to better infer types because it'll recognise it either gets an instance of the $className or null. 

PHPStorm also understands the syntax:

![Scherm­afbeelding 2022-12-11 om 01 52 47](https://user-images.githubusercontent.com/312944/206881355-e83f604c-21ef-4d00-a620-018466a7b654.jpg)


More details:

- https://phpstan.org/blog/generics-in-php-using-phpdocs
- https://phpstan.org/blog/generics-by-examples

Mark-H committed:
Implement generics syntax to allow improved static analysis

wshawn commented:
I "inform" PhpStorm via code like:
```
 /**
     * Retrieves the specified Bible from database and all relationships.
     * @param modX $modX
     * @param string $bibleAbbr The API provided abbreviation.
     * @return Bible
     */
    public function getBibleInfo(modX $modX, string $bibleAbbr): xPDOObject
    {
        $criteria = $modX->newQuery('Sanityllc\Devotional\Model\Bible');
        $criteria->where(['abbr' => $bibleAbbr]);
        return $modX->getObjectGraph(
            'Sanityllc\Devotional\Model\Bible',
            '{"Copyrights":{"Copyright":{"Organizations":{}}}, "Books":{}, "Language":{}}',
            $criteria
        );
    }
```

Mark-H commented:
And with this PR, you wouldn't have to anymore because it will understand the return type automatically! ;) 

I'd also suggest using `Sanityllc\Devotional\Model\Bible::class` instead of the string - just makes it easier to find usages and easier to click through to the class as well. Also worth noting that code you posted will throw a type error when the requested Bible object does not exist, because getObjectGraph may return null instead of the xPDOObject. You could type-hint for `?Bible` instead of `xPDOObject`. 

I did forget to check the Graph methods tho so I'll add it there as well.

Mark-H committed:
Add generics syntax to getObjectGraph and getCollectionGraph as well

opengeek commented:
This is very cool. I'm wondering if it can also be added to the static methods like load()/loadCollection()/etc.

Mark-H commented:
I don't see why not. ;) When adding phpstan to xPDO itself I'm sure those would get flagged. These methods are just the most impactful ones to allow third parties to start benefiting. :)

opengeek committed:
Add generics hints to static load methods

Also adds missing Om\ prefix to @template attributes in xPDO class to properly resolve xPDOObject class
